### PR TITLE
[MLP-160] Integrate newsapi.org API

### PR DESCRIPTION
**Pull request by briansztamfater**
### Background

We need to integrate NewsApi.org API

### Changes done

Created all the endpoints and models, and leaved a short implementation on LoginViewModel (should be completely refactored and decoupled on its respective components and services when @lpelizzaMS rebases his changes)

### Approvers required

@lpelizzaMS @mvazquez-ms @mmaldini33 

### Optional approvers

@jsmuguruza

### CODE REVIEW

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Article.swift [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+//  Created by Brian Sztamfater on 13/4/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import Foundation
+import SwiftyJSON
+
+class Article {
+    
+    var author: String?
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Article.swift]: I'm not sure.. but you should be able to define all these fields as `let` right? Since you're only setting them inside the `init` method.

Also, can you define them as mandatory instead of optional?

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Source.swift [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import Foundation
+import SwiftyJSON
+
+class Source {
+    
+    var id: String!
+    var name: String?
+    var description: String?
+    var url: URL?
+    var language: String?
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Source.swift]: Can you define this as `var language : Languate`  ? (since you already defined the enum)

Also, check if these fields can be defined as `let` too. Such values shouldn't change once instantiated.

File: IOS/SwiftSeedProject/SwiftSeedProject/Info.plist [hunk 3]
```diff
@@ -4,6 +4,8 @@
 <dict>
 	<key>Base URL</key>
 	<string>$(BASE_URL)</string>
+	<key>API Key</key>
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Info.plist]: Can you add something more to this name? Something like `newsapi API key` or ``news API key`

(We can easily have multiple api keys)

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/Extensions/DataRequest+Extensions.swift [hunk 4]
```diff
@@ -30,13 +30,12 @@ extension DataRequest: ApiRequestProtocol {
     }
     
     static func sanitizeError(json: JSON) -> Result<JSON> {
-        if !json["error"].exists() {
+        if json["status"].stringValue == "ok" {
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/Extensions/DataRequest+Extensions.swift:33]: If we want to check for Http Status code, where can we obtain it?

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/Extensions/DataRequest+Extensions.swift [hunk 5]
```diff
@@ -30,13 +30,12 @@ extension DataRequest: ApiRequestProtocol {
...
             return .success(json)
         }
 
-        let code = json["error"]["code"].intValue
-        let message = json["error"]["message"].stringValue
-        let error = NSError(domain: bundleNameKey(), code: code, userInfo: [NSLocalizedDescriptionKey : message])
+        let message = json["message"].stringValue
+        let error = NSError(domain: bundleNameKey(), code: -9999, userInfo: [NSLocalizedDescriptionKey : message])
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/Extensions/DataRequest+Extensions.swift]: Why the -9999? 😄

**Review by mvazquez-ms — COMMENTED** [review]
Just some minor comments..

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/Extensions/DataRequest+Extensions.swift (hunk 4 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/Extensions/DataRequest+Extensions.swift:33]: The closure that calls this function has a reference to the response object, which has a statusCode property. On this method we just parse the error if exists

File: IOS/SwiftSeedProject/SwiftSeedProject/Info.plist (hunk 3 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Info.plist]: Yep!

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Source.swift (hunk 2 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Domain/Source.swift]: Sure, my fault

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/Extensions/DataRequest+Extensions.swift (hunk 5 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/Extensions/DataRequest+Extensions.swift]: This API doesn't return an error code, so I just hardcoded one. I should add an enum for this I think

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/Api.swift [hunk 6]
```diff
@@ -13,13 +13,22 @@ import SwiftyJSON
...
-    private let processUser: (Result<JSON>, (ApiResult<UserM>) -> Void) -> Void = { result, completion in
+    private let processSources: (Result<JSON>, (ApiResult<Array<Source>>) -> Void) -> Void = { result, completion in
         switch result {
-            case .success(let json):
-                let user = UserM(json: json)
-                completion(ApiResult{ return user })
-            case .failure(let error):
-                completion(ApiResult{ throw error })
+        case .success(let json):
+            let sources = json["sources"].arrayValue.map { Source(json: $0) }
```
Comment (briansztamfater) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/Api.swift]: Love this

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Language.swift [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+//  Language.swift
+//  SwiftSeedProject
+//
+//  Created by Brian Sztamfater on 13/4/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import Foundation
+
+enum Language: String {
```
Comment (lpelizzaMS) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Language.swift:11]: Are you sure that these are all languages?

**Review by lpelizzaMS — COMMENTED** [review]
👍

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Language.swift (hunk 7 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Domain/Language.swift:11]: From the API documentation (https://newsapi.org/#apiSources): 

> Possible options: en, de, fr.

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift [hunk 8]
```diff
@@ -7,25 +7,34 @@
...
     public func getAll() -> [T] {
         return persistence.getAll(entityName: T.EntityName)
     }
     
-    public func getEntityBy(identifier: Int) -> T {
+    public func getEntityBy(identifier: String) -> T? {
         return persistence.getBy(entityIdentifier: identifier, entityName: T.EntityName)
     }
     
+    public func createEntity(json: JSON) {
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift]: This is a bit weird..

So, we have

App
'--------'
Domain
'--------'
Infra

`json` belongs to infra, but this class is exposing a method to be used by `App` layer (because public methods are consumed by the upper layer, otherwise they should be protected/private/internal-C#-style). Anyone consuming this method outside a service will be breaking the layered architecture.

On the other hand, it seems you are using it only within Domain layer, and more precisely, inside a child class,, so this method should be protected, and that way we prevent weird scenarios were a ViewModel ends up handling `json`.

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/SourceService.swift [hunk 9]
```diff
@@ -7,10 +7,31 @@
...
+            do {
+                guard let strongSelf = self else {
+                    return
+                }
+                let json = try result.unwrap()
+                let sources = json["sources"].arrayValue
+                sources.forEach { json in
+                    let id = json["id"].stringValue
+                    if let source = strongSelf.getEntityBy(identifier: id) as Source? {
+                        source.updateWithJSON(json: json)
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/SourceService.swift]: Mmm.. about these lines:

`source.updateWithJSON(json: json)`
`strongSelf.createEntity(json: json)`

The call to `updateWithJSON` will update the dabatase right-away? (or only in-memory?) If that's not the case, then maybe we should delegate the update into another class (the same that defines `createEntity`).

Comment (mvazquez-ms) [reply IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/SourceService.swift]: Oh, nevermind, just saw the `strongSelf.save()` call at the end of the loop.

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/ViewModels/SourceListViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+class SourceListViewModel {
+    
+    private let sourceService: SourceService!
+    
+    let sources = MutableObservableArray<SourceViewModel>()
+    
+    init(sourceService: SourceService) {
+        self.sourceService = sourceService
+        self.setupNotifications()
+        self.getSources()
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Application/ViewModels/SourceListViewModel.swift]: It's usually a bad practice to do stuff other than configuration/setup inside the constructor. Calling `getSources()` and `requestSourcesFromServer()` on the constructor can lead to unwanted behavior.

Constructing an object should fail only on exceptional cases (like trying to initialize it with wrong parameters). Here you have a lot of reasons for the constructor to fail (e.g. CoreData, parsing, network, API), the constructor cannot handle any of those error, and you can't expect consumers to handle such errors when instantiating this ViewModel.

Also, you are hiding behavior. Consumer won't expect this class to do requests to an API upon instantiation.

You could do that more evident by adding an extra parameter on the constructor like `preload` or something like that, but I think it's better to provide methods so consumers can perform those actions on demand.

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/DateUtils.swift [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+import Foundation
+
+class DateUtils {
+    
+    struct Format {
+        static let ISO8601 = "yyyy-MM-dd'T'HH:mm:ssZ"
+    }
+    
+    static func parse(dateString: String, format: String) -> Date {
```
Comment (mvazquez-ms) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/DateUtils.swift]: What do you think about using ISO8601 as default value for `format ` parameter?

**Review by mvazquez-ms — COMMENTED** [review]
Just some minor comments, overall looks great!

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/DateUtils.swift (hunk 11 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/DateUtils.swift]: Great idea. Done!

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift [hunk 12]
```diff
@@ -7,25 +7,34 @@
...
+        persistence.add(json: json, entityName: T.EntityName)
+    }
+    
+    public func save() {
+        persistence.save()
```
Comment (lpelizzaMS) [inline IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift]: I think that the service should not has a save method. What do you think?

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift (hunk 8 above)
Comment (lpelizzaMS) [reply IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift]: I think that the Service layer should not be able to save from json. Maybe the api should has some way to save on core data.

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift [hunk 13]
```diff
@@ -7,25 +7,34 @@
...
     
-    private let persistence: Persistence
-    private let restClient: AnyObject? = nil
-    
-    //TODO: we should inclued the API access also
-    init(persistence: Persistence) {
+    private let persistence: Persistence!
+    internal let api: Api!
+
+    init(persistence: Persistence, api: Api) {
```
Comment (lpelizzaMS) [inline IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift]: I think that API is not a good name. I am awful to create names, but maybe @mvazquez-ms  can help us with this.

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift [hunk 14]
```diff
@@ -7,25 +7,34 @@
 //
 
 import Foundation
+import SwiftyJSON
```
Comment (lpelizzaMS) [inline IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift]: I think that Service should not has this kind of imports

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/SourceService.swift (hunk 9 above)
Comment (lpelizzaMS) [reply IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/SourceService.swift]: I agree with marce comment, maybe we should create another class to delegate some logic

**Review by lpelizzaMS — CHANGES_REQUESTED** [review]
it's great!
I think that we should do some little changes before merge. Maybe Marcelo can help us with those changes

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift (hunk 8 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift]: I don't think the network layer should persist objects, because it would be mixing responsibilities and also creating a new dependency on persistance on the Network layer. The service is already delegating the save to the Persistence. What do you think? Maybe I'm not getting something

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift (hunk 12 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift]: Another option would be to make the persistance object protected and call save from the persistance directly. But I wanted to be consistent with the other functions.

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift (hunk 13 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift]: Yep I don't like it either, I bet @mvazquez-ms will come with a good alternative

Comment (mvazquez-ms) [reply IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift]: Yep, `Api` is way too generic. Maybe `RestApiClient`? or just `RestClient`? for the class name.

And that particular instance should be called something like `newsApiRestClient`

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift (hunk 12 above)
Comment (mvazquez-ms) [reply IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift]: > I think that the service should not has a save method. What do you think?

Exposed as a public method? definitely not. Good catch!

> Another option would be to make the persistence object protected and call save from the persistence directly

Yes, I think that's OK. It might look like an inconsistency, but once this services grow and handle all the domain-related stuff, it will make much more sense. Now it looks like a simple persistence proxy

**Review by mmaldini33 — COMMENTED** [review]
I Will wait the changes to review again the code.

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift (hunk 14 above)
Comment (mvazquez-ms) [reply IOS/SwiftSeedProject/SwiftSeedProject/Application/Service/BaseService.swift]: Yep, ideally, not.

In this case, we have 

`internal func createEntity(json: JSON)`

A proper interface for a domain service should be:

`internal func createEntity(entity: T)`

And the same happens with persistence class.. instead of:

`func add(json: JSON, entityName: String)`

we should have

`func add( entity: T, entityName: String)`

JSON usage should be contained within `RectClient` related classes. Once a response is received, it's parsed into an entity or DTO, and then handled by persistence or domain services.

I'm not sure how difficult that could be with CoreData, do you think we can change it?

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/SourceViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    let identifier: Observable<String>
+    let name: Observable<String>
+    let url: Observable<URL>
+    let detail: Observable<String>
+    let language: Observable<Language>
+    let country: Observable<Country>
+    let category: Observable<SourceCategory>
+    
+    init(source: Source) {
+        self.source = source
```
Comment (lpelizzaMS) [inline IOS/SwiftSeedProject/SwiftSeedProject/Application/SourceViewModel.swift]: do we need to have the property source? I am asking because i don't know how works Bond.

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Article+CoreDataProperties.swift [hunk 16]
```diff
@@ -19,8 +19,8 @@ extension Article {
     @NSManaged public var author: String?
     @NSManaged public var title: String?
     @NSManaged public var detail: String?
-    @NSManaged public var url: String?
-    @NSManaged public var urlToImage: String?
+    @NSManaged public var url: URL?
```
Comment (lpelizzaMS) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Article+CoreDataProperties.swift:22]: did you change the dataModel?

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Article.swift [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+//  SwiftSeedProject
+//
+//  Created by Brian Sztamfater on 13/4/17.
+//  Copyright © 2017 Making Sense. All rights reserved.
+//
+
+import Foundation
+import SwiftyJSON
+
+class ArticleM {
```
Comment (lpelizzaMS) [inline IOS/SwiftSeedProject/SwiftSeedProject/Domain/Article.swift]: @bsztamfater-ms  & @mvazquez-ms  We will work with 2 models for each entity?

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+
+import Foundation
+import Alamofire
+import SwiftyJSON
+
+final class RestClient {
+    // MARK: - Private Properties
+    private let manager: ApiManagerProtocol
+    private let persistence: Persistence
+    private let processSources: (Persistence, Result<JSON>, (ApiResult<Array<Source>>) -> Void) -> Void = { persistence, result, completion in
```
Comment (lpelizzaMS) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift:17]: Could we try to create a generic method, couldn't we? What do you think?

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+                }
+                persistence.save()
+                completion(ApiResult{ return articles })
+            case .failure(let error):
+                completion(ApiResult{ throw error })
+        }
+    }
+    
+    // MARK: - Designated Initializer
+    init(manager: ApiManagerProtocol = Configuration.manager, persistence: Persistence) {
```
Comment (lpelizzaMS) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift]: why are you using Configuration.manager on init?

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+        self.manager = manager
+        self.persistence = persistence
+    }
+    
+    // MARK: - Public Methods
+    func getSources(category: SourceCategory? = nil, language: Language? = nil, country: Country? = nil, completion: @escaping (ApiResult<Array<Source>>) -> Void) {
```
Comment (lpelizzaMS) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift:54]: the same question then before, this looks like similar code all the time. but what do you think? maybe I am missing something

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+            guard let strongSelf = self else {
+                return
+            }
+            strongSelf.processArticles(strongSelf.persistence, response.result, completion)
+        }
+    }
+    
+}
+
+struct Configuration {
```
Comment (lpelizzaMS) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift]: why is this struct on the same file than the class RestClient?

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift [hunk 22]
```diff
@@ -20,54 +20,58 @@ class PersistenceData: Persistence {
     }
     
     public func getAll<T: PersistenceObject>(entityName: String) -> [T] {
-        let objectContext = coreDataManager.privateChildManagedObjectContext()
+        let objectContext = coreDataManager.mainManagedObjectContext
```
Comment (lpelizzaMS) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift:23]: why did you change the privateChildManagedObjectContext by MainManagedObjectContext?

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift [hunk 23]
```diff
@@ -20,54 +20,58 @@ class PersistenceData: Persistence {
...
         }
         return []
     }
     
-    public func getBy<T: PersistenceObject>(entityIdentifier: Int, entityName: String) -> T {
-        let objectContext = coreDataManager.privateChildManagedObjectContext()
-        let fetch = NSFetchRequest<NSFetchRequestResult>(entityName: entityName)
-        fetch.predicate = NSPredicate(format: "identifier == %@", entityIdentifier)
+    public func getBy<T: PersistenceObject>(entityIdentifier: String, entityName: String) -> T? {
+        let objectContext = coreDataManager.mainManagedObjectContext
```
Comment (lpelizzaMS) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift:36]: same Q than before, why did you change the child MOC by the main?

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift [hunk 24]
```diff
@@ -20,54 +20,58 @@ class PersistenceData: Persistence {
...
+            return fetchedObjects.count > 0 ? fetchedObjects.item(at: 0) as? T : nil
         } catch let error as NSError {
             fatalError("Failed to fetch \(entityName): \(error)")
-            print("Could not getBy. \(error), \(error.userInfo)")
         }
     }
     
-    public func add(attributes: [String : AnyObject], entityName: String) {
-        let objectContext = coreDataManager.privateChildManagedObjectContext()
+    public func add(attributes: [String : AnyObject]? = nil, entityName: String) -> PersistenceObject {
```
Comment (lpelizzaMS) [inline IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift]: why didn't you use:
func add<T: PersistenceObject>(attributes: [String : AnyObject]? = nil, entityName: String) -> T
instead of :
func add(attributes: [String : AnyObject]? = nil, entityName: String) -> PersistenceObject

**Review by lpelizzaMS — CHANGES_REQUESTED** [review]
It looks great! I wrote some minor comments

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/SourceViewModel.swift (hunk 15 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Application/SourceViewModel.swift]: Not necessary at all, we can remove it. The binded properties are the Observable ones.

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Article.swift (hunk 17 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Domain/Article.swift]: We can remove this class as we are now using Core Data entities. @mvazquez-ms is it ok to delete it?

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift (hunk 18 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift:17]: Yep, but to move on by now I would add a TODO comment, and then refactor this (there are a few thing that we need to improve on RestClient)

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift (hunk 20 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift:54]: As I pointed above, we need a bit of refactor of this class, but I prefer to move on and refactor it when we have more progress on the seed.

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift (hunk 22 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift:23]: Context on new objects was nil after returning the object, so the update was failing

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift (hunk 23 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift:36]: Idem

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift (hunk 24 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift]: Actually I did it that way firstly, but then changed to this one. No specific reason lol. Which advantages and disadvantages we have with the first option over the second one?

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift (hunk 21 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift]: Good point, I forgot to move it to a new file when I first implemented the RestClient. Will do.

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Article+CoreDataProperties.swift (hunk 16 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Domain/Article+CoreDataProperties.swift:22]: Yes, just to use URLs instead of Strings. I marked those attributes as Transformables

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift (hunk 19 above)
Comment (briansztamfater) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift]: We need it to make requests using the Endpoint enum

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift (hunk 24 above)
Comment (lpelizzaMS) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift]: I think that when you use the first one  you don't need to do a casting

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift (hunk 22 above)
Comment (lpelizzaMS) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/PersistenceData.swift:23]: we need to review this because on this new way you are using always the main thread which means that in some cases you are blocking this.

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift (hunk 20 above)
Comment (lpelizzaMS) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift:54]: I totally agree with you that those changes need to be on other PR. But I think that those changes need to create next to this one. we should complete the rest client asap. what do   @mvazquez-ms think?

File: IOS/SwiftSeedProject/SwiftSeedProject/Application/SourceViewModel.swift (hunk 15 above)
Comment (lpelizzaMS) [reply IOS/SwiftSeedProject/SwiftSeedProject/Application/SourceViewModel.swift]: great!

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift (hunk 20 above)
Comment (mmaldini33) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift:54]: I consider that finish the rest client asap is a good idea!

File: IOS/SwiftSeedProject/SwiftSeedProject/Domain/Article.swift (hunk 17 above)
Comment (mvazquez-ms) [reply IOS/SwiftSeedProject/SwiftSeedProject/Domain/Article.swift]: Sure, go ahead, I'm always in favor of removing stuff!

Comment (mmaldini33) [reply IOS/SwiftSeedProject/SwiftSeedProject/Domain/Article.swift]: lol

File: IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift (hunk 20 above)
Comment (mvazquez-ms) [reply IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/Network/RestClient.swift:54]: Yep, let's wrap these changes and create a separate PR for the refactors discussed here.
